src/encoding_binding.cc

```diff
@@ -101,15 +183,67 @@ void BindingData::EncodeInto(const FunctionCallbackInfo<Value>& args) {
   char* write_result = static_cast<char*>(buf->Data()) + dest->ByteOffset();
   size_t dest_length = dest->ByteLength();
 
-  size_t nchars;
-  size_t written = source->WriteUtf8V2(isolate,
-                                       write_result,
-                                       dest_length,
-                                       String::WriteFlags::kReplaceInvalidUtf8,
-                                       &nchars);
+  size_t read = 0;
+  size_t written = 0;
+  v8::String::ValueView view(isolate, source);
+  uint32_t length = view.length();
+
+  if (view.is_one_byte()) {
+    auto data = reinterpret_cast<const char*>(view.data8());
+    simdutf::result result = simdutf::validate_ascii_with_errors(
+        data, std::min(static_cast<size_t>(length), dest_length));
+    written = read = result.count;
+    auto out_addr = write_result;
+    memcpy(out_addr, data, read);
+    out_addr += read;
+    data += read;
+    length -= read;
+    dest_length -= read;
+    if (length != 0 && dest_length != 0) {
+      size_t rest = findBestFit(data, length, dest_length);
+      if (rest != 0) {
+        DCHECK_LE(simdutf::utf8_length_from_latin1(data, rest), dest_length);
+        written += simdutf::convert_latin1_to_utf8(data, rest, out_addr);
+        read += rest;
+      }
+    }
+  } else {
+    auto data = reinterpret_cast<const char16_t*>(view.data16());
+
+    // Check if input has unpaired surrogates - if so, convert to well-formed
+    // first
+    simdutf::result validation_result =
+        simdutf::validate_utf16_with_errors(data, length);
```

Comment:
This isn't great because we are validating the whole string even if the output buffer is tiny.

Perhaps we just need to wait for the new simdutf?

Or we could do simdutf::validate_utf16_with_errors(data, std::min(length, dest_length)).

Reply:
Small footnote: In node.js, simdutf is not a direct dependency but a dependency of v8.